index.d.ts

```diff
@@ -0,0 +1,438 @@
+  finally(): RSVP.Promise<unknown>;
+  then<TResult1 = T, TResult2 = never>(
+    onfulfilled?:
+      | ((value: T) => TResult1 | RSVP.Promise<TResult1>)
+      | undefined
+      | null,
+    onrejected?:
+      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
+      | undefined
+      | null
+  ): RSVP.Promise<TResult1 | TResult2>;
+}
+
+/**
+ * Task instance was canceled before it started.
+ */
+interface InstanceDropped {
+  readonly state: TaskInstanceState.Dropped;
+  readonly hasStarted: false;
+  readonly isCanceled: true;
+  readonly isDropped: true;
+  readonly isFinished: false;
+  readonly isRunning: false;
+  readonly isSuccessful: false;
+  readonly value: undefined;
+  readonly error: Error;
+}
+
+/**
+ * Task instance was canceled before it could finish.
+ */
+interface InstanceCanceled {
+  readonly state: TaskInstanceState.Canceled;
+  readonly hasStarted: true;
+  readonly isCanceled: true;
+  readonly isDropped: false;
+  readonly isFinished: false;
+  readonly isRunning: false;
+  readonly isSuccessful: false;
+  readonly value: undefined;
+  readonly error: Error;
+}
+
+/**
+ * Task instance is currently running, even if it is paused on a yielded
+ * promise.
+ */
+interface InstanceRunning {
+  readonly state: TaskInstanceState.Running;
+  readonly hasStarted: true;
+  readonly isCanceled: false;
+  readonly isDropped: false;
+  readonly isFinished: false;
+  readonly isRunning: true;
+  readonly isSuccessful: false;
+  readonly value: undefined;
+  readonly error: null;
+}
+
+/**
+ * Task instance hasn't begun running yet. Usually because the task is using
+ * the `.enqueue()` task modifier.
+ */
+interface InstanceWaiting {
+  readonly state: TaskInstanceState.Waiting;
+  readonly hasStarted: false;
+  readonly isCanceled: false;
+  readonly isDropped: false;
+  readonly isFinished: false;
+  readonly isRunning: false;
+  readonly isSuccessful: false;
+  readonly value: undefined;
+  readonly error: null;
+}
+
+/**
+ * Task instance ran to completion, but an exception was thrown.
+ */
+interface InstanceError {
+  readonly state: TaskInstanceState.Finished;
+  readonly hasStarted: true;
+  readonly isCanceled: false;
+  readonly isDropped: false;
+  readonly isFinished: true;
+  readonly isRunning: false;
+  readonly isSuccessful: false;
+  readonly value: undefined;
+  readonly error: Error;
+}
+
+/**
+ * Task instance ran to completion successfully, without any exception being
+ * thrown.
+ */
+interface InstanceSuccess {
+  readonly state: TaskInstanceState.Finished;
+  readonly hasStarted: true;
+  readonly isCanceled: false;
+  readonly isDropped: false;
+  readonly isFinished: true;
+  readonly isRunning: false;
+  readonly isSuccessful: true;
+  readonly value: any;
+  readonly error: null;
+}
+
+export type TaskInstance<T> = TaskInstanceBase<T> &
+  (
+    | InstanceDropped
+    | InstanceCanceled
+    | InstanceRunning
+    | InstanceWaiting
+    | InstanceError
+    | InstanceSuccess);
```

Comment:
It’s worth seeing if this works if we drop the magic bits imported for `get`—I haven’t had a chance to test, but I think it *may*, between that and extending from EmberObject? Because at that point it should be actually the same underlying type definition being invoked.

Reply:
Extending from `EmberObject` results in the same error. Removing the union of state interfaces works and also intersection with just one of the state interfaces works. Intersecting with the union of two or more of the states breaks. It looks like `UnwrapComputedPropertyGetter` can't handle unions of interfaces with different signatures, which makes sense given that it uses `infer`.